linux/ibus-keyman/src/test/keymanutil_tests.c

```diff
@@ -1,59 +1,139 @@
+  g_autofree gchar *path = g_strdup_printf("%s%s/%s/", KEYMAN_DCONF_OPTIONS_PATH, TEST_FIXTURE, testname);
+  g_autoptr(GSettings) settings = g_settings_new_with_path(KEYMAN_DCONF_OPTIONS_CHILD_NAME, path);
   g_settings_reset(settings, KEYMAN_DCONF_OPTIONS_KEY);
-  g_object_unref(G_OBJECT(settings));
-  g_free(path);
 }
 
 void
 set_options_key(gchar* testname, gchar** options) {
-  gchar *path = g_strdup_printf("%s%s/%s/", KEYMAN_DCONF_OPTIONS_PATH, TEST_FIXTURE, testname);
-  GSettings *settings = g_settings_new_with_path(KEYMAN_DCONF_OPTIONS_CHILD_NAME, path);
+  g_autofree gchar* path = g_strdup_printf("%s%s/%s/", KEYMAN_DCONF_OPTIONS_PATH, TEST_FIXTURE, testname);
+  g_autoptr(GSettings) settings = g_settings_new_with_path(KEYMAN_DCONF_OPTIONS_CHILD_NAME, path);
   g_settings_set_strv(settings, KEYMAN_DCONF_OPTIONS_KEY, (const gchar* const*)options);
-  g_object_unref(G_OBJECT(settings));
-  g_free(path);
 }
 
 gchar**
 get_options_key(gchar* testname) {
-  gchar* path = g_strdup_printf("%s%s/%s/", KEYMAN_DCONF_OPTIONS_PATH, TEST_FIXTURE, testname);
-  GSettings* settings = g_settings_new_with_path(KEYMAN_DCONF_OPTIONS_CHILD_NAME, path);
+  g_autofree gchar* path = g_strdup_printf("%s%s/%s/", KEYMAN_DCONF_OPTIONS_PATH, TEST_FIXTURE, testname);
+  g_autoptr(GSettings) settings = g_settings_new_with_path(KEYMAN_DCONF_OPTIONS_CHILD_NAME, path);
   gchar** result = g_settings_get_strv(settings, KEYMAN_DCONF_OPTIONS_KEY);
-  g_object_unref(G_OBJECT(settings));
-  g_free(path);
   return result;
 }
 
+kmp_keyboard*
+_get_kmp_keyboard(gchar* version, gchar** languages) {
+  kmp_keyboard* keyboard = g_new0(kmp_keyboard, 1);
+  keyboard->name         = g_strdup("Testing");
+  keyboard->id           = g_strdup("tst");
+  keyboard->version      = g_strdup(version);
+  keyboard->kmx_file     = g_strdup("tst.kmx");
+  keyboard->kvk_file     = g_strdup("");
+  keyboard->languages    = NULL;
+  for (gchar* lang = *languages++; lang; lang = *languages++) {
+    gchar** tokens         = g_strsplit(lang, ":", 2);
+    kmp_language* kmp_lang = g_new0(kmp_language, 1);
+    kmp_lang->id           = g_strdup(tokens[0]);
+    kmp_lang->name         = g_strdup(tokens[1]);
+    g_strfreev(tokens);
+    keyboard->languages = g_list_append(keyboard->languages, kmp_lang);
+  }
+  return keyboard;
+}
+
+kmp_info*
+_get_kmp_info(gchar * copyright, gchar * author_desc, gchar * author_url) {
+  kmp_info* info    = g_new0(kmp_info, 1);
+  info->copyright   = g_strdup(copyright);
+  info->author_desc = g_strdup(author_desc);
+  info->author_url  = g_strdup(author_url);
+  return info;
+}
+
+keyboard_details*
+_get_keyboard_details(gchar * description, gchar * license) {
+  keyboard_details* details = g_new0(keyboard_details, 1);
+  details->id               = g_strdup("tst");
+  details->description      = g_strdup(description);
+  details->license          = g_strdup(license);
+  return details;
+}
+
+add_keyboard_data*
+_get_keyboard_data() {
```

Comment:
Ah, initially I wondered why _get_kmp_foo() and _get_keyboard_foo() weren't in kmpdetails.c, but I see these generate test stubs. Since the keyboard ID is `tst`, I suggest `_get_tst_` prefixes.

To me, that would make it a little clearer some of the info isn't coming from kmp.json. (ll 36-39)

Reply:
I couldn't think of a name that would make it clearer, but using `tst` is a good suggestion.